### PR TITLE
For #41415: Adds tests!

### PR DESCRIPTION
**Pull request by thebeeland**
Implements a basic testing framework based on tk-core, and implements a round of tests for api_v2. Also included is a small bug fix. The issue was that immutable configs didn't get shotgun_xxx.yml entity types added to the entity type whitelist, which was incorrect behavior.

### CODE REVIEW

File: tests/api_v2/test_cache.py [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+
+        # These should not match, because one Task entity is linked to a
+        # Shot entity and the other to an Asset.
+        self.assertNotEqual(key_5, key_6)
+
+    def test_contents_hash(self):
+        """
+        Tests to ensure that the contents hash is properly constructed out of
+        the component parts: a config descriptor, and the env yml file modtimes
+        in the same where a mutable config is in use.
```
Comment (jfboismenu) [inline tests/api_v2/test_cache.py]: is the same

File: tests/api_v2/test_cache.py [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+        # changing the way it would on a page refresh or navigation.
+        self.api._cache = dict()
+
+        hash_6 = self.api._get_contents_hash(
+            config_descriptor,
+            self.api._get_software_entities(),
+        )
+
+        self.assertNotEqual(hash_5, hash_6)
+
```
Comment (jfboismenu) [inline tests/api_v2/test_cache.py]: Hound-CI would not be pleased. Lots of empty lines.

File: tests/api_v2/test_util_methods.py [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+# This work is provided "AS IS" and subject to the Shotgun Pipeline Toolkit 
+# Source Code License included in this distribution package. See LICENSE.
+# By accessing, using, copying or modifying this work you indicate your 
+# agreement to the Shotgun Pipeline Toolkit Source Code License. All rights 
+# not expressly granted therein are reserved by Shotgun Software Inc.
+
+import os
+import sys
+
+from tank_test.tank_test_base import *
```
Comment (jfboismenu) [inline tests/api_v2/test_util_methods.py]: Do not `import *`. This prevents linters from discovering unused variables. Only import what you need, which is `setUpModule` in this case.

File: tests/api_v2/test_util_methods.py [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+            actions,
+            self.api._get_software_entities(),
+            project,
+        )
+
+        # Make sure one got filtered out, and that the one remaining is
+        # what's expected.
+        self.assertEqual(len(filtered_actions), 1)
+        self.assertEqual(filtered_actions[0], actions[0])
+
```
Comment (jfboismenu) [inline tests/api_v2/test_util_methods.py]: Woof!

File: tests/fixtures/config/bundles/test_app/info.yml [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+# Source Code License included in this distribution package. See LICENSE.
+# By accessing, using, copying or modifying this work you indicate your 
+# agreement to the Shotgun Pipeline Toolkit Source Code License. All rights 
+# not expressly granted therein are reserved by Shotgun Software Inc.
+
+# Metadata defining the behaviour and requirements for this engine
+
+# expected fields in the configuration file for this engine
+configuration:
+    debug_logging:
```
Comment (jfboismenu) [inline tests/fixtures/config/bundles/test_app/info.yml]: Probably not used, right?

File: tests/fixtures/config/bundles/test_app/info.yml [hunk 6]
```diff
@@ -0,0 +1,25 @@
+# Copyright (c) 2013 Shotgun Software Inc.
```
Comment (jfboismenu) [inline tests/fixtures/config/bundles/test_app/info.yml]: 2017

File: tests/fixtures/config/bundles/test_app/app.py [hunk 7]
```diff
@@ -0,0 +1,22 @@
+# Copyright (c) 2013 Shotgun Software Inc.
```
Comment (jfboismenu) [inline tests/fixtures/config/bundles/test_app/app.py]: 2017

File: tests/fixtures/config/bundles/test_engine/engine.py [hunk 8]
```diff
@@ -0,0 +1,39 @@
+# Copyright (c) 2013 Shotgun Software Inc.
```
Comment (jfboismenu) [inline tests/fixtures/config/bundles/test_engine/engine.py]: 2017

File: tests/fixtures/config/bundles/test_engine/engine.py [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+
+class TestEngine(Engine):
+    
+    def init_engine(self):
+        pass
+                
+    ##########################################################################################
+    # logging interfaces
+
+    def log_debug(self, msg):
```
Comment (jfboismenu) [inline tests/fixtures/config/bundles/test_engine/engine.py]: base class implements logging now.

File: tests/fixtures/config/bundles/test_engine/info.yml [hunk 10]
```diff
@@ -0,0 +1,22 @@
+# Copyright (c) 2013 Shotgun Software Inc.
```
Comment (jfboismenu) [inline tests/fixtures/config/bundles/test_engine/info.yml]: 2017

File: tests/fixtures/config/bundles/test_engine/info.yml [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+# Source Code License included in this distribution package. See LICENSE.
+# By accessing, using, copying or modifying this work you indicate your 
+# agreement to the Shotgun Pipeline Toolkit Source Code License. All rights 
+# not expressly granted therein are reserved by Shotgun Software Inc.
+
+# Metadata defining the behaviour and requirements for this engine
+
+# expected fields in the configuration file for this engine
+configuration:
+    debug_logging:
```
Comment (jfboismenu) [inline tests/fixtures/config/bundles/test_engine/info.yml]: Same comment

File: tests/fixtures/config/env/test.yml [hunk 12]
```diff
@@ -0,0 +1,16 @@
+engines:
+
+    test_engine:
+
+        location: {'type': 'dev', 'path': '{PIPELINE_CONFIG}/config/bundles/test_engine'}
+        debug_logging: false
```
Comment (jfboismenu) [inline tests/fixtures/config/env/test.yml]: you'll be able to remove the setting.

File: tests/python/base_test.py [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+# This work is provided "AS IS" and subject to the Shotgun Pipeline Toolkit 
+# Source Code License included in this distribution package. See LICENSE.
+# By accessing, using, copying or modifying this work you indicate your 
+# agreement to the Shotgun Pipeline Toolkit Source Code License. All rights 
+# not expressly granted therein are reserved by Shotgun Software Inc.
+
+import os
+import sys
+
+from tank_test.tank_test_base import *
```
Comment (jfboismenu) [inline tests/python/base_test.py]: Import TankTestBase and setUpModule only.

File: tests/python/base_test.py [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+import sgtk
+from mock_host import MockHost
```
Comment (jfboismenu) [inline tests/python/base_test.py]: Maybe we should repackage all of these modules under something like
```
from integration_tests import MockHost
from integration_tests import MockDescriptor
from integration_tests import TankDesktopServerFramework
```
Having a bunch of absolute includes for stuff that is not meant to be shared worries me a bit.

File: tests/python/base_test.py [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+        Fixtures teardown
+        """
+        # engine is held as global, so must be destroyed.
+        cur_engine = sgtk.platform.current_engine()
+        if cur_engine:
+            cur_engine.destroy()
+
+        # important to call base class so it can clean up memory
+        super(TestDesktopServerFramework, self).tearDown()
+
```
Comment (jfboismenu) [inline tests/python/base_test.py]: Woof!

File: tests/python/mock_descriptor.py [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+
+    def get_path(self):
+        return self._path
+
+    def is_immutable(self):
+        return self._is_immutable
+
+    def get_uri(self):
+        return self._uri
+
```
Comment (jfboismenu) [inline tests/python/mock_descriptor.py]: Woof!

**Comment by thebeeland** [comment]
The failing builds is a problem with the VM the tests run on not having libffi5.

**Pull request merged**